### PR TITLE
feat: add `math/base/special/cfloorf`

### PR DESCRIPTION
**Pull request by aayush0325**
Resolves #649.

## Description

> What is the purpose of this pull request?

This pull request:

-   Resolves part of #649 by adding `math/base/special/cfloorf`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Open to reviews!!

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by aayush0325** [comment]
Good Afternoon, can i kindly get your inputs in this @gunjjoshi ?

File: lib/node_modules/@stdlib/math/base/special/cfloorf/README.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# Floor
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/README.md]:
```suggestion
# cfloorf
```
Let's match this with what we have in `cceilf`: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/cceilf/README.md?plain=1#L21

File: lib/node_modules/@stdlib/math/base/special/cfloorf/README.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+<section class="examples">
+
+## Examples
+
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var Complex64 = require( '@stdlib/complex/float32/ctor' );
+var randu = require( '@stdlib/random/base/randu' );
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/README.md]:
```suggestion
var uniform = require( '@stdlib/random/base/uniform' ).factory;
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/README.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+var w;
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+    re = ( randu()*100.0 ) - 50.0;
+    im = ( randu()*100.0 ) - 50.0;
+    z = new Complex64( re, im );
+    w = cfloorf( z );
+    console.log( 'cfloorf(%s) = %s', z.toString(), w.toString() );
+}
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/README.md]:
```suggestion
var rand = uniform( -50.0, 50.0 );

var z;
var i;
for ( i = 0; i < 100; i++ ) {
    z = new Complex64( rand(), rand() );
    console.log( 'cfloorf(%s) = %s', z, cfloorf( z ) );
}
```

**Comment by aayush0325** [comment]
have pushed the changes @gunjjoshi , thanks!

File: lib/node_modules/@stdlib/math/base/special/cfloorf/README.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+#include "stdlib/complex/float32/ctor.h"
+#include "stdlib/complex/float32/reim.h"
+#include <stdio.h>
+
+int main( void ) {
+    const stdlib_complex64_t x[] = {
+        stdlib_complex128( 3.14f, 1.5f ),
+        stdlib_complex128( -3.14f, -1.5f ),
+        stdlib_complex128( 0.0f, 0.0f ),
+        stdlib_complex128( 0.0f/0.0f, 0.0f/0.0f )
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/README.md]:
```suggestion
        stdlib_complex128( 0.0f / 0.0f, 0.0f / 0.0f )
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/benchmark.native.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	var i;
+
+	values = [
+		new Complex64( uniform( -500.0, 500.0 ), uniform( -500.0, 500.0 ) ),
+		new Complex64( uniform( -500.0, 500.0 ), uniform( -500.0, 500.0 ) )
+	];
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = cfloorf( values[ i%values.length ] );
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/benchmark.native.js]:
```suggestion
		y = cfloorf( values[ i % values.length ] );
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/Makefile [hunk 6]
```diff
@@ -0,0 +1,107 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2018 The Stdlib Authors.
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/Makefile]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/benchmark.c [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+	float complex z;
+	float complex y;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		re = ( 1000.0f * rand_float() ) - 500.0f;
+		im = ( 1000.0f * rand_float() ) - 500.0f;
+		z = re + im*I;
+		y = floorf( crealf(z) ) + floorf( cimagf(z) )*I;
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/benchmark.c]:
```suggestion
		y = floorf( crealf( z ) ) + floorf( cimagf( z ) )*I;
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/binding.gyp [hunk 8]
```diff
@@ -0,0 +1,170 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2023 The Stdlib Authors.
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/binding.gyp]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/examples/c/Makefile [hunk 9]
```diff
@@ -0,0 +1,146 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2023 The Stdlib Authors.
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/examples/c/Makefile]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/examples/c/example.c [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+#include "stdlib/complex/float32/ctor.h"
+#include "stdlib/complex/float32/reim.h"
+#include <stdio.h>
+
+int main( void ) {
+	const stdlib_complex64_t x[] = {
+		stdlib_complex64( 3.14f, 1.5f ),
+		stdlib_complex64( -3.14f, -1.5f ),
+		stdlib_complex64( 0.0f, 0.0f ),
+		stdlib_complex64( 0.0f/0.0f, 0.0f/0.0f )
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/examples/c/example.c]:
```suggestion
		stdlib_complex64( 0.0f / 0.0f, 0.0f / 0.0f )
```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/examples/index.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+var w;
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+	re = ( randu()*100.0 ) - 50.0;
+	im = ( randu()*100.0 ) - 50.0;
+	z = new Complex64( re, im );
+	w = cfloorf( z );
+	console.log( 'cfloorf(%s) = %s', z.toString(), w.toString() );
+}
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/examples/index.js]: We should make the same changes here as well, which we made at https://github.com/stdlib-js/stdlib/pull/3058#discussion_r1826939884 and https://github.com/stdlib-js/stdlib/pull/3058#discussion_r1826941117.

File: lib/node_modules/@stdlib/math/base/special/cfloorf/include.gypi [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2023 The Stdlib Authors.
```
Comment (gunjjoshi) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/include.gypi]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

**Comment by aayush0325** [comment]
pushed the changes @gunjjoshi .

File: lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/benchmark.c [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+	float complex z;
+	float complex y;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		re = ( 1000.0f * rand_float() ) - 500.0f;
+		im = ( 1000.0f * rand_float() ) - 500.0f;
+		z = re + im*I;
+		y = floorf( crealf( z ) ) + floorf( cimagf( z ) )*I;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/benchmark/c/benchmark.c]: The C benchmark code manually computes the floor of the real and imaginary parts using `floorf`:

 ```c
 y = floorf( crealf( z ) ) + floorf( cimagf( z ) )*I;
 ```

 I would have expected a call to the `stdlib_base_cfloorf` function instead:

 ```c
 y = stdlib_base_cfloorf( z );
 ```

File: lib/node_modules/@stdlib/math/base/special/cfloorf/README.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+#include "stdlib/complex/float32/ctor.h"
+#include "stdlib/complex/float32/reim.h"
+#include <stdio.h>
+
+int main( void ) {
+    const stdlib_complex64_t x[] = {
+        stdlib_complex128( 3.14f, 1.5f ),
+        stdlib_complex128( -3.14f, -1.5f ),
+        stdlib_complex128( 0.0f, 0.0f ),
+        stdlib_complex128( 0.0f / 0.0f, 0.0f / 0.0f )
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/cfloorf/README.md]:
```suggestion
        stdlib_complex64( 3.14f, 1.5f ),
        stdlib_complex64( -3.14f, -1.5f ),
        stdlib_complex64( 0.0f, 0.0f ),
        stdlib_complex64( 0.0f / 0.0f, 0.0f / 0.0f )
```

**Review by Planeshifter — COMMENTED** [review]
Should be close to land (left two comments), thanks for your work on this!

**Comment by aayush0325** [comment]
wasn't exactly sure on how to proceed with this but i tried making some changes to make the benchmark work with `stdlib_base_cfloorf`, kindly let me know if i should take another approach @Planeshifter

**Review by Planeshifter — APPROVED** [review]
Did another review pass; looks good to me. Thank you, @aayush0325!

**Pull request merged**